### ISSUE TITLE
React Router works locally but not on Heroku

### ISSUE THREAD

**Issue by rosscournoyer602** [mars/create-react-app-buildpack#98]
When I am running on localhost, I can navigate around starting at root, and my history works properly, and the same is true on Heroku. Also, on localhost, I can enter any route and pull up a specific page or entry. This requires the component to look at the window.pathname, check if a displayed entry is on the Redux state, and if it is not, make an API call to get it and put it on state. This does not work once deployed to Heroku. I am using create-react-app, not ejected, with Redux, and making API calls using Redux Sagas.

`Uncaught SyntaxError: Unexpected token <` in main.*****.js line one

For some reason, it is not expecting html, and is being served the index.html file, hence the unexpected <

I'm stumped by how vague this error is, so I'm not sure what to show you. Router is set up at top level component here:

```
ReactDOM.render(
    <Provider store={store}>
      <Router>
        <App />
      </Router>
    </Provider>
, document.getElementById('root'))
```

Nothing crazy I think, and I have a static.json file like this:

```
{
    "root":"build/",
    "clean_urls": false,
    "routes": {
        "/**": "index.html"
    }
}

```
Sorry for the lack of specificity here, I am just not sure how to diagnose the problem. Others have fixed it with static.json, but that doesn't seem to be working for me. Thanks!

**Comment by rosscournoyer602** [comment]
For now, I seem to have fixed this by using HashRouter instead of BrowserRouter. Sorry for frivolous issue. Thanks for a very awesome and useful buildpack!

**Comment by mars** [comment]
Hi @rosscournoyer602 ,

The `static.json` config that is known to work is [documented](https://github.com/mars/create-react-app-buildpack#user-content-routing-clean-urls). It does not contain the `"clean_urls": false` entry like yours. Maybe that's breaking it.

Also, did you deploy the `static.json` changes?

```bash
git add static.json
git commit -m 'Update static routing'
git push heroku master
```

**Comment by stefroussos** [comment]
Hi @mars ,

I'm getting the same `Uncaught SyntaxError: Unexpected token <` error as @rosscournoyer602 , but would like to keep using BrowserRouter instead of HashRouter. 
My `index.js`
```jsx
ReactDOM.render(
  <Provider store={store}>
    <BrowserRouter>
      <Switch>
        <Route exact path="/" component={Home} />
        <Route path="/app" component={App} />
        <Route path="/signin" component={SignIn} />
        <Route path="/signup" component={SignUp} />
      </Switch>
    </BrowserRouter>
  </Provider>, document.getElementById('root')
);
```
and `static.json`
```json
{
  "root": "build/",
  "routes": {
    "/**": "index.html"
  }
}
```
any changes to the `static.json` file result in a `404`

**Comment by brycepavey** [comment]
@stefroussos did you find a solution with keeping BrowserRouter? we still have this happen on every deploy we do and can't track it down

**Comment by stefroussos** [comment]
@brycepavey I was using github pages for testing, which messed up the buildpack. After I deleted `gh-pages` and fixed my `package.json`, everything worked well.

**Comment by etaytch** [comment]
Having the same issue :(

**Comment by mars** [comment]
Where does `Uncaught SyntaxError: Unexpected token <` appear? Is it during deploy/build on Heroku?

Please paste additional context for the error, like the build log or app logs.

Does running `npm run build` locally throw the same error?

**Comment by jhbed** [comment]
I am also having this issue. Router works perfectly locally, but when pushing to heroku the router does not work at all. I've tried all the solutions above but to no avail. 

index.js
```javascript
ReactDOM.render(
  <BrowserRouter>
  <ApolloProvider client={client}>
    <App />
  </ApolloProvider>
  </BrowserRouter>,
  
  document.getElementById('root')
)
```

App.js
```javascript
class App extends Component {
  render() {
    return (
      <div>
        <Header />
        <Switch>
          <Route exact path="/" component={EventCardList} />
          <Route exact path="/create" component={CreateEventForm} /> 
        </Switch>
      </div>
    );
  }
}
```

static.json
```json
{
    "root": "build/",
    "routes": {
      "/**": "index.html"
    }
}
```

Thank you for the help!

**Comment by etaytch** [comment]
Hey @mars , sorry for the late response 🙏 
The issue occurred only when deploying to Heroku.
Locally it worked just fine. 
I was able to solve this by changing the `homepage` attribute in `package.json` to my Heroku url 😨 .

**Comment by jhbed** [comment]
Thank you for the update @etaytch - unfortunately this solution did not work for me.

**Comment by mikeyyyzhao** [comment]
We're also experiencing a similar issue. Our webpage works fine if you go to /home and then navigate from there (./home/example...). However, on refresh, (./home/example..) errors out and returns Uncaught SyntaxError: Unexpected token <. It also errors out if you type in ./home/anything_else but if you type in ./anything_else  (where home used to be) it works. Our code works fine on all these cases locally. I've tried the suggestions above and still stuck on this. Any help/suggestions would be greatly appreciated!

Here's our static.json file 
```
{
  "root": "build/",
  "routes": {
    "/home/**": "index.html",
    "/**": "index.html"
  }
}
```

and our App.js
```
render() {
    return (
      <Router>
        <div className="App">
          <Switch>
            <Route path="/login" component={Login} />

            <AuthorizedRoute path="*" component={HomePage} />
          </Switch>
        </div>
      </Router>
    );
  }
```

Basically, if the user types in /login, we route them to login but for anything other route, we check if they're logged in, and if they are, we take them to home otherwise, redirect them to login.

**Comment by NikolSkvarilova** [comment]
Hello,

I have the same problem, but I only get an err 404. I have changed my `static.json` according to the @mikeyyyzhao's post. 

My `app.js`:

```jsx
<Router>
    <div className="App">
        <Switch>
            <Route path="/" exact component={Home} />
            <Route path="/play/:level" component={Play} />
         </Switch>
    </div>
</Router>
```

My `static.json`:
```json
{
  "root": "build/",
  "routes": {
      "/**": "index.html",
      "/play/**": "index.html"
  }
}
```

Log from Heroku:
```
2020-07-27T19:57:02.965973+00:00 heroku[router]: at=info method=GET path="/play/3" host=not-just-sudoku.herokuapp.com request_id=78cf8243-c915-4896-a986-75dd6e247b72 fwd="46.149.119.135" dyno=web.1 connect=1ms service=3ms status=404 bytes=400 protocol=https
2020-07-27T19:57:02.966620+00:00 app[web.1]: 10.97.137.170 - - [27/Jul/2020:19:57:02 +0000] "GET /play/3 HTTP/1.1" 404 232 "https://not-just-sudoku.herokuapp.com/" "Mozilla/5.0 (X11; Linux x86_64; rv:79.0) Gecko/20100101 Firefox/79.0"
```

Obviously I get an error when I try to go to the `/play/some_int`.

**Comment by jorgenfaunaapp** [comment]
I to get the 'Uncaught SyntaxError: Unexpected token <' error when trying to load my page across different devices. Using hashrouter solved it on some platforms, but on mobile my site still will not load.

My app.js
```
function App() {
  return (
    <ApolloProvider client={apolloClient}>
      <ThemeProvider theme={theme}>
        <LocalizationProvider locale={"dk"}  dateAdapter={LuxonUtils}>
        <AlertContainer/>
            <AppWrapperContainer authenticated={authenticated()}/>
        </LocalizationProvider>
      </ThemeProvider>
      <TawkTo/>
    </ApolloProvider>
  );
}
```

My AppWrapperContainer
```
export const AppWrapperContainer = (props) => {
    return (<Router>
        <Switch>
            {unprotectedRoutes.map((route, key) => (
                <UnprotectedRoute key={key} {...route} />
            ))}
          <Route component={AuthenticatedRoutes} path={routeUrls.authenticated} />
          <LoginContainer path={routeUrls.login}/>
          <Redirect path={"/"} to={{pathname: homeUrl}}/>
          <Redirect to={{pathname: routeUrls.login}}/>
        </Switch>
    </Router>)
}

```
static.json

```
{
  "root": "build/",
  "routes": {
    "/*": "index.html",
    "/**": "index.html"
  },
  "https_only": true
} 
```
